### ISSUE TITLE
False positive: unused-public-var with #(fn) and macroexpand hook

### ISSUE THREAD

**Issue by mrkam2** [clojure-lsp/clojure-lsp#1766]
**Describe the bug**
clojure-lsp/unused-public-var reports a function as unused when it is used in an anonymous function literal inside a macro configured with a hook.

`a.clj`:
(ns a)

(defn my-fn [x] x)

(defmacro my-do [& body]
  `(do ~@body))

(defn -main []
  ; Leaving only this line will trigger the unused public var message.
  (my-do (map #(my-fn %) (range 10))))
```

**To Reproduce**
1 . Unzip [issue.zip](https://github.com/clojure-lsp/clojure-lsp/files/14148694/issue.zip)
2. Run `clojure-lsp diagnostics`
3. See error
src/a.clj:3:7: info: [clojure-lsp/unused-public-var] Unused public var 'a/my-fn'

**Expected behavior**
No such error should be reported.

When either `:analyze-call` hook is used, or `#(my-fn %)` is replaced with `(fn [x] (my-fn [x]))`, unused variable is not reported.

**Comment by borkdude** [comment]
The var is available in the analysis but has a `:derived-location` because `#(...)` cannot be represented as an s-expression, so when going through macro-expansion the position is derived. Probably lsp ignores this usage, but that's not clj-kondo's issue.

**Comment by ericdallo** [comment]
Yeah, we ignore :derived-location otherwise we will cause a breaking change on custom hooks, this was implemented [here](https://github.com/clojure-lsp/clojure-lsp/commit/b032d6873824ccaf5c2bc6d76d0d5dc13d98f553), so not sure how to fix that in clojure-lsp without breaking custom hooks that was why we ignored `derived-location`.

**Comment by borkdude** [comment]
You don't need a location to know that a var is used, even if the var usage has a derived location.
